src/physics/base/ImportedProcessAdapter.cc

```diff
@@ -0,0 +1,205 @@
+                             particles,
+                             std::move(process_class),
+                             {pdg_numbers.begin(), pdg_numbers.end()})
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the interaction cross sections for the given material and particle.
+ */
+auto ImportedProcessAdapter::step_limits(Applicability range) const
+    -> StepLimitBuilders
+{
+    CELER_EXPECT(ids_.count(range.particle));
+    CELER_EXPECT(range.material);
+
+    // Get list of physics tables
+    const ParticleProcessIds& ids = ids_.find(range.particle)->second;
+    const ImportProcess&      import_process = imported_->get(ids.process);
+
+    auto get_vector = [&range, &import_process](ImportTableId table_id) {
+        CELER_ASSERT(table_id < import_process.tables.size());
+        const ImportPhysicsTable& tab = import_process.tables[table_id.get()];
+        CELER_ASSERT(range.material < tab.physics_vectors.size());
+        return tab.physics_vectors[range.material.get()];
+    };
+
+    StepLimitBuilders builders;
+
+    // Construct cross section tables
+    if (ids.lambda && ids.lambda_prim)
+    {
+        // Both unscaled and scaled values are present
+        const auto& lo = get_vector(ids.lambda);
+        CELER_ASSERT(lo.vector_type == ImportPhysicsVectorType::log);
+        const auto& hi = get_vector(ids.lambda_prim);
+        CELER_ASSERT(hi.vector_type == ImportPhysicsVectorType::log);
+        builders[ValueGridType::macro_xs] = ValueGridXsBuilder::from_geant(
+            make_span(lo.x), make_span(lo.y), make_span(hi.x), make_span(hi.y));
+    }
+    else if (ids.lambda_prim)
+    {
+        // Only high-energy (energy-scale) cross sections are presesnt
+        const auto& vec = get_vector(ids.lambda_prim);
+        CELER_ASSERT(vec.vector_type == ImportPhysicsVectorType::log);
+        builders[ValueGridType::macro_xs] = ValueGridXsBuilder::from_scaled(
+            make_span(vec.x), make_span(vec.y));
+    }
+    else if (ids.lambda)
+    {
+        // Only low-energy cross sections are presesnt
+        const auto& vec = get_vector(ids.lambda);
+        CELER_ASSERT(vec.vector_type == ImportPhysicsVectorType::log);
+        builders[ValueGridType::macro_xs] = ValueGridLogBuilder::from_geant(
+            make_span(vec.x), make_span(vec.y));
+    }
+
+    // Construct slowing-down data
+    if (ids.dedx)
+    {
+        const auto& vec = get_vector(ids.dedx);
+        CELER_ASSERT(vec.vector_type == ImportPhysicsVectorType::log);
+        builders[ValueGridType::energy_loss] = ValueGridLogBuilder::from_geant(
+            make_span(vec.x), make_span(vec.y));
+    }
+
+    // Construct range limiters
+    if (ids.range)
+    {
+        const auto& vec = get_vector(ids.range);
+        CELER_ASSERT(vec.vector_type == ImportPhysicsVectorType::log);
+        builders[ValueGridType::range] = ValueGridLogBuilder::from_geant(
```

Comment:
I see that there is a `ValueGridRangeBuilder` for range tables, but it does not seem to be fully implemented. Is there any particular issue in not using it here? Range tables are also log(E), so probably not, but it was worth mentioning.

Reply:
Oh good catch -- I'd meant to use it. It's fully implemented but all it does is call the regular log builder with an extra check that the ranges are monotonically increasing.